src/Host/Broker/Impl/Lifetime/LifetimeManager.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.R.Host.Broker.Startup;
+
+namespace Microsoft.R.Host.Broker.Lifetime {
+    public class LifetimeManager {
+        private readonly LifetimeOptions _options;
+        private readonly ILogger _logger;
+
+        private CancellationTokenSource _cts;
+
+        public LifetimeManager(IOptions<LifetimeOptions> options, ILogger<LifetimeManager> logger) {
+            _options = options.Value;
+            _logger = logger;
+        }
+
+        public void Initialize() {
+            if (_options.ParentProcessID != null) {
+                int pid = _options.ParentProcessID.Value;
+                Process process;
+                try {
+                    process = Process.GetProcessById(pid);
+                    process.EnableRaisingEvents = true;
+                } catch (ArgumentException) {
+                    _logger.LogCritical($"Designated parent process {pid} not found");
+                    Program.Exit();
+                    return;
+                }
+
+                _logger.LogInformation($"Monitoring parent process {pid}");
+                process.Exited += delegate {
+                    _logger.LogInformation($"Parent process {pid} exited, shutting down.");
+                    Program.Exit();
+                };
+            }
+
+            Ping();
+        }
+
+        public void Ping() {
```

Comment:
Yep - but only if the timeout value was actually specified (currently the local connector doesn't use it, and neither connector actually sends pings).

Not sure if we will actually need it for lifetime purposes at all. For local scenario, having broker monitor VS process proved to be just as reliable, and doesn't pollute the logs with constant pings. Ping itself is still useful as a quick way to verify if server is up or not, though.